src/components/editProfessorsModal/hooks/useEditProfessorsModal.ts

```diff
@@ -0,0 +1,177 @@
+import { SelectChangeEvent } from '@mui/material';
+import { useEffect, useMemo, useState } from 'react';
+import useAssignProfessorsRequest from '../../../service/requests/useAssignProfessorsRequest';
+import {
+  TCompleteSubject,
+  TSubjectResponsible,
+} from '../../../service/requests/useGetSubject/types';
+import useListProfessorsRequest from '../../../service/requests/useListProfessorsRequest';
+import { useSnackBar } from '../../../utils/renderSnackBar';
+import { TSelectedProfessor } from './types';
+import useEndSubjectResponsability from '../../../service/requests/useEndSubjectResponsability';
+
+type Props = {
+  refetchSubject(): void;
+};
+
+const useEditProfessorsModal = ({ refetchSubject }: Props) => {
+  const { data: listProfessorsResponse, listProfessors } =
+    useListProfessorsRequest();
+  const { isSuccess, isLoading, assignProfessors, error, resetStates } =
+    useAssignProfessorsRequest();
+  const {
+    isSuccess: isSuccessEndResponsability,
+    isLoading: isLoadingEndResponsability,
+    endResponsability,
+    error: errorEndResponsability,
+    resetStates: resetStatesResponsability,
+  } = useEndSubjectResponsability();
+
+  const { showErrorSnackBar, showSuccessSnackBar } = useSnackBar();
+
+  const [isOpen, setIsOpen] = useState(false);
+  const [isRemoveProfessorModalOpen, setIsRemoveProfessorModalOpen] =
+    useState(false);
+  const [selectedProfessorRemove, setSelectedProfessorRemove] =
+    useState<TSubjectResponsible>();
+  const [selectedSubject, setSelectedSubject] = useState<TCompleteSubject>();
+  const [selectedProfessorsIds, setSelectedProfessorsIds] = useState<string[]>(
+    []
+  );
+
+  const subjectReponsibles: TSubjectResponsible[] = useMemo(() => {
+    if (!listProfessorsResponse || !selectedSubject) return [];
+    return selectedSubject.responsables.filter(
+      (responsible) => responsible.status.status === 'Aprovado'
+    );
+  }, [listProfessorsResponse, selectedSubject]);
+
+  const professors: TSelectedProfessor[] = useMemo(() => {
+    if (!listProfessorsResponse || !selectedSubject) return [];
+
+    const notAssignedProfessors = listProfessorsResponse.data.filter(
+      (prof) =>
+        prof.SubjectResponsability.find(
+          (resp) =>
+            resp.subject.code === selectedSubject?.code &&
+            resp.status.status === 'Aprovado'
+        ) === undefined
```

Comment:
O problema que vc disse é porque você tá fazendo esse memo orientado à lista de professores, porém como ela não muda, quando vc remove ou add uma responsabilidade essa lista não percebe a alteração. O correto é o seguinte:

```suggestion
        !selectedSubject.responsables.find(
          (resp) =>
            resp.professor.user.id === prof.user.id &&
            resp.status.status === 'Aprovado'
        )
```

Assim, quando a disciplina selecionada for recarregada e já trouxer a lista de responsáveis atualizada, vc vai se basear nela e não na lista de todos os professores, cuja responsabilidade não tá atualizada.

Reply:
boa, troquei aqueer